### PR TITLE
KAFKA-5279: TransactionCoordinator must expire transactionalIds

### PR DESCRIPTION
**Pull request by dguy**
remove transactions that have not been updated for at least `transactional.id.expiration.ms`

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang @hachikuji i suspect this will have to change a bit once https://github.com/apache/kafka/pull/3082 is merged

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (apurvam) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: I am a bit confused about the locking here. We seem to already hold the `stateLock` when this method is called. Are we guaranteed that the call back is scheduled on the same thread? Otherwise, we will have a deadlock.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (apurvam) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Should this be on error level? Is it going to be retried indefinitely?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (apurvam) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Should this be `DefaultRemoveExpiredTransactionalIdsIntervalMs`?

**Review by apurvam — COMMENTED** [review]
Looks good to me overall. The only meaningful question/comment is about the use of the `stateLock` lock.

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: nit: `...ExpiredTransactionalId...`

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: We should only expire transactional id if it does not have ongoing txns, i.e. the state should only be empty (just initialized the pid but never started any txns) or completeXX (previous txns has been completed).

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: We can call `time` outside the synchronization block.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Thinking about it a bit more, we do not need to whole the lock when appending txn messages to local logs as we discussed in another PR: https://github.com/apache/kafka/pull/3082#discussion_r117588659

Plus, holding the lock here may result in a dead lock situation, since in purgatory we hold the delayed producer object when calling `tryComplete` which will call the `removeFromCacheCallback` below. So there can be thread 1 trying to grab `statelock` first then the delayed producer `object` lock later in `tryCompleteElseWatch`, while thread 2 trying to grab the `object` lock first in `tryComplete` when an fetch request is handled and hence HW has been advanced, and then inside `removeFromCacheCallback` trying to grab the `statelock`, hence the deadlock.

We can consider 1) only grab the lock while getting the list of txnIds to expire along with its belonged partitions, 2) release the lock and then try to append to log.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Maybe `config.requestTimeoutMs.toLong`? Default is 30 secs anyways.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: We need to check the coordinator epoch to make sure it matches, otherwise these entries may be from a new epoch and hence should not be removed.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: I am concerning if the following can happen:

1) tc decides to expire a txnId, and have written the tombstone.
2) an `addPartitions` is received from the txnId, and is accepted and handled since no checks will fail.
3) tombstone is replicated, the cache is removed from the cache.

Should we guard the object if we have decided to remove the txnId? For example, we can add a new state, say `Dead` in the `TransactionState`, and set its pending state to that before we decided to expire and append to local log. And when handling `addPartition` and `endTxn`, if its pending state is `Dead` return concurrent txn to let it retry later, and when handling `initPid` and pending state is `Dead`, bump up the producer epoch as we did; then on the `removeFromCacheCallback`, check the pending state / coordinator epoch / state to make sure it is `Dead / same as the old epoch / same as the old state` and only remove the cache if it passes.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Should we also set the `delay` to `transactionalIdExpirationMs` otherwise the first task will be scheduled immediately and we will have nothing to expire.

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala]: Related to the above comment, let's test for the cases that after we have decided to expire a txnId, another `addPartition` is received and handled.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala]: nit: change the scehduled task name to `scheduler.schedule("transaction-abort"...)` as well?

**Review by guozhangwang — COMMENTED** [review]
Thanks for the patch. Left some comments.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: I don't think we will be in the lock when the callback happens, i.e., `replicaManager.appendRecords(...)` will return once it has written to the local log. The callback will happen once all replicas have ack'd

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: It can be error level i guess. Yes, it will be retried when the scheduler next runs

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: @guozhangwang i thought the lock would be released as soon as the `replicaManager.appendRecords(..)` had returned from writing to the local log. But i guess you are saying the callback could happen immediately? Anyway, makes sense to release the lock before the log append. Thanks

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Yeah - i wasn't sure about this. But i guess it makes sense as any timed out transactions will have been rolled back, so will be aborted.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: actually needs to be `config.removeExpiredTransactionalIdsIntervalMs`

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: For `initPid` if a pending state is set it already will return  `Errors.CONCURRENT_TRANSACTION` which should suffice? I don't see any reason to make it more complicated than that.
I'm not sure about the `addPartition` and `endTxn` cases as how would you get into the state where a Transaction is likely dead and then you try to `addPartition` or `endTxn`? i.e., the transaction will have been aborted long before then if it timed out. So in the `addPartition` case then transaction will have had to have been previously `Empty` in which case it has already been through `initPid` and backed off if there is a pending state defined.

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala
Comment (dguy) [reply core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala]: I don't see how this is possible. You'd have to first go through `initPid` which we will return a retriable exceptions for as the metadata would be in a pending state.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (apurvam) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: @dguy, I agree about the `addPartition` and `endTxn` cases. We should not get those requests if we are in a position to expire a transactional Id, but yet, we may. If the client suddenly starts a new transaction, it may send a valid `addPartitions` request. 

In general, doesn't there seem to be a race here? If there is no open transaction for the producer, and we decide to remove the transactional id, we could be in the process of removing it while a new transaction starts, or a new producer with the same transactionalId comes online.

To be specific, what prevents this scenario: 

0. TransactionalId `foobar` is picked for expiration.
1. An `addPartitions` with transactionalId `foobar` comes in. There is no pending state and at this point the transactionalId is in the cache, so the transaction is started. We may even update the cache with the partitions and `OnGoing` state.
2. The `append` to delete the transactionalId is written to the log, and the cache is cleared. 
4. The ongoing transaction is now corrupt. 

This may happen right? If not, what prevents it?

Comment (guozhangwang) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: @dguy By `Dead` I meant to indicate that the txnId has been dead, not the ongoing transaction since it should have long-been aborted or completed. The problem is that when we are writing the tombstone into the txn log we did not set any pending state in the txn metadata object itself to fence any future request on this txnId while the append is being replicated. So it is possible that after we have written the tombstone but the delayed produce has not been completed, an `addPartition` is received with the txnId.

BTW I think if the current ongoing transaction of the transactionId has complete committed or aborted, its metadata's state should be in CompleteCommit or CompleteAbort, and `Empty` should only happen when this txnId has only been registered from an `initPid` but no txns have ever been started yet, i.e. no `addPartitions` have been received yet.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:186]: Actually I was wrong before: we do need to hold the state lock to make sure the retrieval of the metadata object from the cache and the append call is in the same monitoring block: https://github.com/apache/kafka/pull/3082/files#r117879684

Note that we need to make sure that when we grab locks, we should always grab the statelock for the whole metadata map cache, and then the lock on the metadata object itself. So in the `removeFromCacheCallback` we cannot grab the lock on statelock since we are already grabbing the lock on the metadata object itself in `delayedOperation.tryComplete()`, and this out-of-ordering will cause dead-lock.

Instead, we can just take the `readlock` after #3082 so that this out-or-ordering will be Okay, note that `metadataPerTransactionalId` is a `Pool` which is thread-safe so `readlock` is sufficient.

Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:186]: Ok - i think i'll just wait until 3082 is merged before going any further

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: @apurvam @guozhangwang i think it is ok in this case now. I've made the change to put the metadata into a pending state of `Dead`. So if an `addPartitions` came in after this has happened, it will result in `Errors.CONCURRENT_TRANSACTIONS`

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: since we are not grabbing a lock on the object, we should make `@volatile txnLastUpdateTimestamp` to make sure we get the latest value.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (apurvam) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: So to be clear, if we get an `AddPartitions` when we are in a `Dead` state, the coordinator will return a `CONCURRENT_TRANSACTIONS` error. The client will retry after a back off. 

After the transactional Id is fully removed, the next retry of `AddPartitions` will hit an `InvalidPidMapping` since there would be no record of the transactional Id in the cache. At that point, it will error out. Correct?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala
Comment (apurvam) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala]: In these and other messages, we should include all the transaction metadata, including transcationalId, producerId, producerEpoch, and the precise from and to state transition which is invalid.

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala
Comment (dguy) [reply core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala]: scratch that it is possible. I keep forgetting that you don't need to `initPid` for each transaction - doh!

File: core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala]: Hmm, it seems Dead is only possible in the `state` if there is a race that someone access the cache in between the two consecutive commands in the `removeFromCacheCallback`, i.e. after the `completeTransitionTo` call but before the `remove` call. In that case I'm wondering if it is OK to just remove it after checking the pending state matches `Dead` instead of really overwrite the `state`?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala]: nit: extra space.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala]: Ditto above.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala]: is it intentional?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala]: Ditto above, and plus if we cannot skip setting the state to `dead` we can merge these two cases as `Empty | Dead`.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala]: Would you just want to modify line 321 below?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Could we log it if the epoch does not match and hence the operation is aborted?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Do we really want to log it as `error`? I think both `txn loading in progress` and `not coordinator` are normal to observe.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:193]: We can omit `None` here since it is the default value.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: See comments above; also if we have to call `completeTransitionTo` and it did throw the exception,  should we clear the `pendingState` to None then, otherwise all future requests will be fenced?

**Review by guozhangwang — COMMENTED** [review]
Left a few more comments.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:193]: I know,  but i needed it for the tests otherwise EasyMock won't match

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: i had it as debug, think apruva suggested error. what should it be?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala]: I'm not following what you are suggesting. Yes i think there is a chance of a race in that the thread here gets the metadata from the cache when it is in `pendingState` of `Dead`, but before this thread synchronizes on the `txnMetadata` the `removeFromCacheCallback` completes the  transition to `Dead`

Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala]: ok - i think you are saying leave it in `pendingState` of `Dead` and remove it from the cache?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala
Comment (dguy) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala]: Yeah i should have! Will remove anyway as we aren't going to transition to `Dead`

**Review by apurvam — APPROVED** [review]
LGTM!

**Comment by guozhangwang** [comment]
LGTM! Merged to trunk and 0.11.0.

**Pull request closed without merging**
